File: sub_pub/core/interfaces.py
"""Abstract base classes for message sources and publishers"""
from abc import ABC, abstractmethod
from typing import AsyncIterator, Iterator, Optional
from sub_pub.core.message import Message


class MessageSource(ABC):
    """Abstract base class for message sources"""
    
    @abstractmethod
    def connect(self) -> None:
        """Establish connection to the message source"""
        pass
    
    @abstractmethod
    def subscribe(self, topics: list[str]) -> None:
        """Subscribe to one or more topics"""
        pass
    
    @abstractmethod
    def consume(self) -> Iterator[Message]:
        """Consume messages from subscribed topics"""
        pass
    
    @abstractmethod
    def close(self) -> None:
        """Close the connection to the message source"""
        pass
    
    @abstractmethod
    def commit(self, message: Optional[Message] = None) -> None:
        """Commit message offset (for systems that support it)"""
        pass


class MessagePublisher(ABC):
    """Abstract base class for message publishers"""
    
    @abstractmethod
    def connect(self) -> None:
        """Establish connection to the message destination"""
        pass
    
    @abstractmethod
    def publish(self, message: Message, topic: str) -> None:
        """Publish a message to a topic"""
        pass
    
    @abstractmethod
    def flush(self) -> None:
        """Flush any buffered messages"""
        pass
    
    @abstractmethod
    def close(self) -> None:
        """Close the connection to the message destination"""
        pass


class AsyncMessageSource(ABC):
    """Abstract base class for async message sources"""

    @abstractmethod
    async def connect(self) -> None:
        """Establish connection to the message source"""
        pass

    @abstractmethod
    async def subscribe(self, topics: list[str]) -> None:
        """Subscribe to one or more topics"""
        pass

    @abstractmethod
    async def consume(self) -> AsyncIterator[Message]:
        """Consume messages from subscribed topics"""
        pass

    @abstractmethod
    async def close(self) -> None:
        """Close the connection to the message source"""
        pass

    @abstractmethod
    async def commit(self, message: Optional[Message] = None) -> None:
        """Commit message offset (for systems that support it)"""
        pass


class AsyncMessagePublisher(ABC):
    """Abstract base class for async message publishers"""

    @abstractmethod
    async def connect(self) -> None:
        """Establish connection to the message destination"""
        pass

    @abstractmethod
    async def publish(self, message: Message, topic: str) -> None:
        """Publish a message to a topic"""
        pass

    @abstractmethod
    async def flush(self) -> None:
        """Flush any buffered messages"""
        pass

    @abstractmethod
    async def close(self) -> None:
        """Close the connection to the message destination"""
        pass

File: rust/src/core/interfaces.rs
use async_trait::async_trait;

use crate::core::message::Message;
use crate::error::SubPubError;

/// Trait for anything that can produce messages (a message source / consumer).
///
/// Implementations are responsible for connecting to their backing system,
/// subscribing to topics, and delivering messages one at a time via `consume`.
#[async_trait]
pub trait MessageSource: Send + Sync {
    /// Establish a connection to the backing message system.
    async fn connect(&mut self) -> Result<(), SubPubError>;

    /// Subscribe to the given list of topics.
    async fn subscribe(&mut self, topics: &[&str]) -> Result<(), SubPubError>;

    /// Fetch the next available message.  Returns `None` when the source is
    /// exhausted or closed.
    async fn consume(&mut self) -> Result<Option<Message>, SubPubError>;

    /// Acknowledge / commit the given message so it won't be re-delivered.
    async fn commit(&mut self, message: &Message) -> Result<(), SubPubError>;

    /// Cleanly shut down this source.
    async fn close(&mut self) -> Result<(), SubPubError>;
}

/// Trait for anything that can accept and forward messages (a publisher).
#[async_trait]
pub trait MessagePublisher: Send + Sync {
    /// Establish a connection to the backing message system.
    async fn connect(&mut self) -> Result<(), SubPubError>;

    /// Publish a single message to `topic`.
    async fn publish(&mut self, message: &Message, topic: &str) -> Result<(), SubPubError>;

    /// Flush any buffered messages, ensuring they are delivered.
    async fn flush(&mut self) -> Result<(), SubPubError>;

    /// Cleanly shut down this publisher.
    async fn close(&mut self) -> Result<(), SubPubError>;
}
